### PR TITLE
DDP-4276 implement copy handlers

### PR DESCRIPTION
**Pull request by yufengwng**
## Context and the big picture

This PR implements the actual copying functionality, which processes through the copy configuration and copy answers from source location to the appropriate target. Both the copy-answer and create-instance event actions now hook into this "executor" to do the copying. This implementation is feature-parity with old copying functionality, so existing studies should continue to work in the same way.
 
 ## What type of change is this?
  
  - [ ] Bugfix (non-breaking change which fixes an issue)
  - [x] New feature (non-breaking change which adds functionality)
  - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - [ ] Experimental proof-of-concept
  - [ ] Infosec/Compliance remediation
  
 ## Risk Assessment
  - [x] These changes are **HIGH RISK**.
    - [x] There is an elevated risk that may impact a large number of features
    - [x] There is an elevated risk to security and privacy (since we're copying answers)
    - [ ] These changes introduce new 3rd party dependencies, alter the network perimeter, or contain
    major upgrades to persistent storage or SaaS dependencies such as auth0, sendgrid, or easypost    
  - [ ] These changes are low risk, do not impact security/privacy, do not impact pepper shared
  components/modules, and have no major upgrades to 3rd party services or libraries
 
 ### FUD Score 
 Overall, how are you feeling about these changes?
  - [ ] :relaxed: All good, business as usual!
  - [x] :sweat_smile: There might be some issues here (the composite answers are tough...)
  - [ ] :scream:I'm sweaty and nervous
 
 ## How to we demo these changes?
 How does one observe these changes in a deployed system?  Note that _user visible_ encompasses
 many personas--not just patients and study staff, but ops your fellow devs, compliance, etc.
 
 - [x] They are user-visible in dev as a regular user journey and require no additional instructions.
 - [ ] Getting dev into a state where this is user-visible requires some tech fiddling.  I have
 documented these steps in the related ticket.
 - [ ] Requires other features before it's human visible.  I have documented the blocking issues
 in jira.
 - [ ] I have no idea how to demo this.  Please help me!
 
 ## UI/UX
  - [x] There ain't no UI/UX in these changes
  - [ ] My changes have passed muster with Erin and Jen via an over-the-shoulder review, screenshots, etc.
  - [ ] Erin and Jen have not had an opportunity to provide feedback yet
 
###  Backend
   - [x] I have considered error handling and notification to slack alert rooms via `LOG.error()`.
     #### Route Changes
     - [ ] My changes involve a new route changes to an existing route, so the who/what/when are logged using the usual logback mechanism
at the _start_ of the route, prior to any validation, as well as at the end of the route when work completes.
     ### Housekeeping Changes
     - [ ] My changes involve updates to housekeeping, and I am logging the who/what/when via logback.
  ### Client-side
   - [ ] My changes do not involve backend route change, but I have useful logging statements nonetheless.

## Analytics
 - [ ] I have discussed the analytics needs at both a platform and study level with Jen and instrumented code
 accordingly.
 - [x] There are no analytics requirements for these changes

## Security and Privacy
Ensuring proper security and privacy for our users is essential.

  ###  Backend
   - [ ] These changes alter fundamental auth filter logic
   - [ ] These changes introduce a new backend API route that is behind an auth filter
   - [ ] These changes introduce a new backend API that _does not require auth_
   - [ ] These changes alter auth0 internals (rules, hosted login, configuration, etc.)
   - [ ] These changes expose new information to elastic and I have verified that PHI is not being exposed to the public
    
  ### Browser
  - [ ] These changes alter what we store in browser storage
  
  ### Mobile
  - [ ] TBD 
 
 ## Performance and Stability
  - [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, and performance bottlenecks and written a brief summary in this PR
  - [ ] I'm unsure about stability, performance, fault tolerance, and graceful degradation.  Please help!
  
## Testing
 - [x] I have written automated positive tests
 - [x] I have written automated negative tests
 - [ ] I have written zero automated tests but have poked around locally to verify proper functionality
 - [ ] The jira ticket has acceptance criteria and Kiara has in the information she needs to test my changes
 
## Release
 - [x] These changes require no special release procedures--just code!
 - [ ] Releasing these changes requires special handling
   - [ ] I have documented the special procedures in the release plan document
   - [ ] The special procedures require minimal downtime
   - [ ] The special procedures require unavoidable, extended downtime for:
     - [ ] Participant UX
     - [ ] Study staff

### CODE REVIEW

**Review by zyme — COMMENTED** [review]
Since we have the original test, could we run that with both old and new copiers and get a sense of performance?  If there's a big difference there, we should think about what that might mean for user experience.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/copy/AnswerToAnswerCopier.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger(AnswerToAnswerCopier.class);
+
+    private final Handle handle;
+    private final long operatorId;
+    private final JdbiCompositeQuestion jdbiCompositeQuestion;
+    private final JdbiCompositeAnswer jdbiCompositeAnswer;
+    private final AnswerDao answerDao;
+
+    public AnswerToAnswerCopier(Handle handle, long operatorId) {
```
Comment (zyme) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/copy/AnswerToAnswerCopier.java]: Maybe `operatorId` -> `operatorIdForNewAnswer`?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/event/CopyAnswerEventAction.java [hunk 2]
```diff
@@ -1,106 +1,47 @@
...
     public CopyAnswerEventAction(EventConfiguration eventConfiguration, EventConfigurationDto dto) {
         super(eventConfiguration, dto);
         copyConfigurationId = dto.getCopyActionCopyConfigurationId();
-        answerDao = AnswerDao.fromSqlConfig(ConfigFactory.load(ConfigFile.SQL_CONFIG_FILE));
     }
 
     @Override
-    public void doAction(PexInterpreter pexInterpreter, Handle handle, EventSignal eventSignal) {
+    public void doAction(PexInterpreter interpreter, Handle handle, EventSignal signal) {
         // FIXME The Copy Answer Action was originally written to assume that the activity that triggered the copy
```
Comment (zyme) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/event/CopyAnswerEventAction.java]: Are these fixmes relevant?

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/ActivityInstanceSql.sql.stg [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  left join date_answer as da on da.answer_id = ans.answer_id
+  left join numeric_answer as na on na.answer_id = ans.answer_id
+  left join picklist_option__answer as pa on pa.answer_id = ans.answer_id
+  left join picklist_option as po on po.picklist_option_id = pa.picklist_option_id
+  left join composite_answer_item as ca on ca.parent_answer_id = ans.answer_id
+>>
+
+// Sort by participant, then instance, then composite child answers.
+all_form_instance_responses_with_latest_status_and_answers_order_by() ::= <<
+order by ai.participant_id asc,
```
Comment (zyme) [inline pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/ActivityInstanceSql.sql.stg:74]: Could you say a few words about the sorting here?  I'm not seeing where the sort order is leveraged.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/event/CopyAnswerEventAction.java (hunk 2 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/model/event/CopyAnswerEventAction.java]: Removed fixmes after finishing implementation.

File: pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/ActivityInstanceSql.sql.stg (hunk 3 above)
Comment (yufengwng) [reply pepper-apis/src/main/resources/org/broadinstitute/ddp/db/dao/ActivityInstanceSql.sql.stg:74]: Expanded on comment to explain sorting.

**Comment by yufengwng** [comment]
please run tests

**Comment by yufengwng** [comment]
@zyme the original copier and this one are very different implementations, so I'm afraid the comparison might not be fair or realistic. Nonetheless, I tried the following: I swapped out some things in `AnswerToAnswerCopier` with the original copier sql, timed the loop in `CopyExecutor` and used `ActivityInstanceCreationEventActionTest` as the driver.

My limited observations show that the two implementations seem comparable, with this new implementation being slightly faster (e.g. it usually takes new impl. < 100ms to do all the copy pairs in the test, while original impl. > 100ms). Again, it doesn't paint the full picture since handling of composite answers are very different. And I suspect the time savings in new impl. are mostly due to the fact that we have all the needed source answers in memory already whereas the original impl. is doing queries for each one.

**Pull request merged**